### PR TITLE
[Feature] Add Retrieval Augmented Generation search processors

### PR DESCRIPTION
**Pull request by austintlee**
### Description
#1150 
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ x] New functionality includes testing.
  - [x ] All tests pass
- [ x] New functionality has been documented.
  - [ x] New functionality has javadoc added
- [ x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by austintlee** [comment]
```
curl -X PUT http://localhost:9200/_cluster/settings -H "Content-Type: application/json" \
-d '{"persistent": {"plugins.ml_commons.rag_pipeline_feature_enabled" : true}}' | jq .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   190  100   116  100    74   1830   1168 --:--:-- --:--:-- --:--:--  3454
{
  "acknowledged": true,
  "persistent": {
    "plugins": {
      "ml_commons": {
        "rag_pipeline_feature_enabled": "true"
      }
    }
  },
  "transient": {}
}
```

```
[2023-09-01T13:59:54,088][INFO ][o.o.c.s.ClusterSettings  ] [integTest-0] updating [plugins.ml_commons.rag_pipeline_feature_enabled] from [false] to [true]
```

**Comment by austintlee** [comment]
@ylwu-amzn

**Comment by austintlee** [comment]
The failing test(s) seem to be flaky ones.  I ran the last one that failed on my machine and it passed.

File: plugin/src/main/java/org/opensearch/ml/settings/MLCommonsSettings.java
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/settings/MLCommonsSettings.java:134]: I'm curious to know what would actually happen if someone were to enable the `ML_COMMONS_RAG_PIPELINE_FEATURE_ENABLED` flag but not enable the `ML_COMMONS_MEMORY_FEATURE_ENABLED` flag, or vice versa?

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java
Comment (dhrubo-os) [inline search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java:137]: do we need this line?

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/client/ConversationalMemoryClient.java
Comment (dhrubo-os) [inline search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/client/ConversationalMemoryClient.java:74]: [minor] `allPagesFetched`?

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/prompt/PromptUtil.java
Comment (dhrubo-os) [inline search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/prompt/PromptUtil.java:43]: I remember you said you did some experiments on this prompt template to get the better result from LLM. I'm just curious to know more about that experiments. If you can share any doc/reference, that'll be helpful.

**Comment by dhrubo-os** [comment]
> The failing test(s) seem to be flaky ones. I ran the last one that failed on my machine and it passed.

Seems like it's a jarhell issue? 

```
Task :opensearch-ml-plugin:jarHell FAILED
```

File: plugin/src/main/java/org/opensearch/ml/settings/MLCommonsSettings.java
Comment (austintlee) [reply plugin/src/main/java/org/opensearch/ml/settings/MLCommonsSettings.java:134]: "Memory" is the most fundamental piece and it can be used independently by any application or a REST client.  You CAN use RAG without memory as well if you are not passing a conversationId.  I think we need to make sure we document these use cases and scenarios as part of the 2.10.0 release.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/prompt/PromptUtil.java
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/prompt/PromptUtil.java:43]: Ack.  @HenryL27 can provide more details on this (but perhaps outside this PR?)

Comment (dhrubo-os) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/prompt/PromptUtil.java:43]: > 

Yeah, sure. That's not going to block this PR.

File: plugin/src/main/java/org/opensearch/ml/settings/MLCommonsSettings.java
Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/settings/MLCommonsSettings.java:134]: @HenryL27 could you please explain `Trying to use memory anyway with that disabled will throw the feature flag error.`  --> I assume customer will not see this error?

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java:137]: This is used to construct a single string out of an array as a helper method to store the search results as a "additionalInfo" string in memory.

**Comment by austintlee** [comment]
@dhrubo-os 

```
| Exception in thread "main" java.lang.IllegalStateException: jar hell!
| class: org.bouncycastle.LICENSE
| jar1: /Users/taehe/.gradle/caches/modules-2/files-2.1/org.bouncycastle/bcprov-jdk15to18/1.75/df22e1b6a9f6b218913f5b68dd16641344397fe0/bcprov-jdk15to18-1.75.jar
| jar2: /Users/taehe/.gradle/caches/modules-2/files-2.1/org.bouncycastle/bcprov-ext-jdk15on/1.70/373d425c5ecb4edc9e3e2f7f7ff39bc8eff4abbf/bcprov-ext-jdk15on-1.70.jar
|       at org.opensearch.bootstrap.JarHell.checkClass(JarHell.java:316)
|       at org.opensearch.bootstrap.JarHell.checkJarHell(JarHell.java:215)
|       at org.opensearch.bootstrap.JarHell.checkJarHell(JarHell.java:102)
|       at org.opensearch.bootstrap.JarHell.main(JarHell.java:86)
```

This seems unrelated to our change, though.

**Comment by HenryL27** [comment]
> @dhrubo-os
> 
> ```
> | Exception in thread "main" java.lang.IllegalStateException: jar hell!
> | class: org.bouncycastle.LICENSE
> | jar1: /Users/taehe/.gradle/caches/modules-2/files-2.1/org.bouncycastle/bcprov-jdk15to18/1.75/df22e1b6a9f6b218913f5b68dd16641344397fe0/bcprov-jdk15to18-1.75.jar
> | jar2: /Users/taehe/.gradle/caches/modules-2/files-2.1/org.bouncycastle/bcprov-ext-jdk15on/1.70/373d425c5ecb4edc9e3e2f7f7ff39bc8eff4abbf/bcprov-ext-jdk15on-1.70.jar
> |       at org.opensearch.bootstrap.JarHell.checkClass(JarHell.java:316)
> |       at org.opensearch.bootstrap.JarHell.checkJarHell(JarHell.java:215)
> |       at org.opensearch.bootstrap.JarHell.checkJarHell(JarHell.java:102)
> |       at org.opensearch.bootstrap.JarHell.main(JarHell.java:86)
> ```
> 
> This seems unrelated to our change, though.

This is also broken in 2.x

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/client/ConversationalMemoryClient.java
Comment (HenryL27) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/client/ConversationalMemoryClient.java:68]: https://github.com/opensearch-project/ml-commons/issues/1268

File: plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java
Comment (austintlee) [reply plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java:683]: This list and the other two below are registered at plugin load time and are used to populate immutable collections.  This feature flag is not tied to a REST API or a Transport action so it needs to be enabled before the plugin is enabled.

https://github.com/opensearch-project/OpenSearch/blob/aca2e9de7daaadee77c33e21ff5215c3f1e8600f/server/src/main/java/org/opensearch/search/SearchModule.java#L1068

https://github.com/opensearch-project/OpenSearch/blob/aca2e9de7daaadee77c33e21ff5215c3f1e8600f/server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java#L115

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/client/ConversationalMemoryClient.java
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/client/ConversationalMemoryClient.java:45]: For 2.10, let's keep it internal, but we can discuss making it external for 2.11?

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/client/ConversationalMemoryClient.java
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/client/ConversationalMemoryClient.java:68]: Yeah, let's do that for 2.11.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/client/ConversationalMemoryClient.java
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/client/ConversationalMemoryClient.java:80]: OK!

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/client/ConversationalMemoryClient.java
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/client/ConversationalMemoryClient.java:75]: Will add a check for it.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/client/ConversationalMemoryClient.java
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/client/ConversationalMemoryClient.java:54]: Done!

File: settings.gradle
Comment (austintlee) [inline settings.gradle:19]: @dhrubo-os ^^

**Comment by dhrubo-os** [comment]
> This is also broken in 2.x

Yeah, core merged a [PR](https://github.com/opensearch-project/OpenSearch/pull/9289/files). Seems like this is causing the issue.

File: plugin/src/main/java/org/opensearch/ml/settings/MLCommonsSettings.java
Comment (austintlee) [reply plugin/src/main/java/org/opensearch/ml/settings/MLCommonsSettings.java:134]: @dhrubo-os this is the error we throw:

```
new OpenSearchException(
                        "The experimental Conversation Memory feature is not enabled. To enable, please update the setting "
                            + ConversationalIndexConstants.ML_COMMONS_MEMORY_FEATURE_ENABLED.getKey()

```

**Comment by austintlee** [comment]
A fix (#1281) is now available.  Should I cherry pick it once it gets approved?

**Comment by austintlee** [comment]
@dhrubo-os @jngz-es besides the jarHell issue (which we should first fix in 2.x), is there anything remaining on this PR?

File: plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java
Comment (jngz-es) [reply plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java:683]: Sorry for the confusion. What I meant is that in the future, if we want to add another ext with another flag, we need to remove this return statement and add another two statements with flags. For example,

List<> list = new ArrayList();
if (flagA) { // add ext A in list}
if (flagB) { // add ext B in list}
return list

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/client/ConversationalMemoryClient.java
Comment (jngz-es) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/client/ConversationalMemoryClient.java:46]: sounds good

File: plugin/src/main/java/org/opensearch/ml/settings/MLCommonsSettings.java
Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/settings/MLCommonsSettings.java:134]: Sorry for the confusion. I'm trying to understand it step by step:

1. I enabled `ML_COMMONS_RAG_PIPELINE_FEATURE_ENABLED` 
2. I asked a question in the RAG pipeline now. Can I continue my Q/A interaction with LLM now? Or it will show me this `OpenSearchException` to enable the other feature flag?

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/prompt/PromptUtil.java
Comment (jngz-es) [inline search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/prompt/PromptUtil.java:62]: Can we rename Role to like ChatRole? I think it is specific for chat. The name "Role" is too general which could be confusing to the security role.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/client/ConversationalMemoryClient.java
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/client/ConversationalMemoryClient.java]: Ok!

File: plugin/src/main/java/org/opensearch/ml/settings/MLCommonsSettings.java
Comment (austintlee) [reply plugin/src/main/java/org/opensearch/ml/settings/MLCommonsSettings.java:134]: It won't.  You will only get the error if you pass it a conversionId.

File: plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java
Comment (austintlee) [reply plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java]: Got it.

File: plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java
Comment (austintlee) [reply plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java]: Ok

File: plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java
Comment (austintlee) [reply plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java]: Ok.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/prompt/PromptUtil.java
Comment (dhrubo-os) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/prompt/PromptUtil.java:62]: @austintlee @HenryL27 do we have plan to address this?

Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/prompt/PromptUtil.java:62]: Sure.

Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/prompt/PromptUtil.java:62]: Just to clear, this enum is internal to PromptUtil, but I will rename it to ChatRole just to make it clearer.

**Review by dhrubo-os — APPROVED** [review]
Thanks for adding unit tests.

**Review by jngz-es — APPROVED** [review]

**Pull request merged**
